### PR TITLE
Quest Editor implemented

### PR DESCRIPTION
**Pull request by mquellhorst**
Implemented DraftJS for a more comprehensive description of quests

paired with: @luxumbra

### CODE REVIEW

**Comment by mquellhorst** [comment]
When editing a quest the app end up in an infinite loop because the html-to-draft package isn't usable in typescript.
Working on a solution

**Comment by mquellhorst** [comment]
Would be nice if someone can have a look.
To convert the draftjs state to html for storage we need the draftjs-to-html package.
For the creation of quests this is working fine but somehow it messes up the edit page.
Here is the error message.

```
@metafam/web: [0] ReferenceError: document is not defined
@metafam/web: [0]     at getSafeBodyFromHTML (/home/mquellhorst/Workspace/MetaGame/thegame/node_modules/draft-js/lib/getSafeBodyFromHTML.js:25:19)
@metafam/web: [0]     at convertFromHTMLToContentBlocks (/home/mquellhorst/Workspace/MetaGame/thegame/node_modules/draft-js/lib/convertFromHTMLToContentBlocks.js:773:18)
@metafam/web: [0]     at getDefaultFormValues (/home/mquellhorst/Workspace/MetaGame/thegame/packages/web/.next/server/pages/quest/[id]/edit.js:6743:92)
@metafam/web: [0]     at /home/mquellhorst/Workspace/MetaGame/thegame/packages/web/.next/server/pages/quest/[id]/edit.js:6824:85
@metafam/web: [0]     at Object.useMemo (/home/mquellhorst/Workspace/MetaGame/thegame/node_modules/react-dom/cjs/react-dom-server.node.development.js:1239:19)
@metafam/web: [0]     at useMemo (/home/mquellhorst/Workspace/MetaGame/thegame/node_modules/react/cjs/react.development.js:1521:21)
@metafam/web: [0]     at QuestForm (/home/mquellhorst/Workspace/MetaGame/thegame/packages/web/.next/server/pages/quest/[id]/edit.js:6824:78)
@metafam/web: [0]     at processChild (/home/mquellhorst/Workspace/MetaGame/thegame/node_modules/react-dom/cjs/react-dom-server.node.development.js:3043:14)
@metafam/web: [0]     at resolve (/home/mquellhorst/Workspace/MetaGame/thegame/node_modules/react-dom/cjs/react-dom-server.node.development.js:2960:5)
@metafam/web: [0]     at ReactDOMServerRenderer.render (/home/mquellhorst/Workspace/MetaGame/thegame/node_modules/react-dom/cjs/react-dom-server.node.development.js:3435:22)
@metafam/web: [0]     at ReactDOMServerRenderer.read (/home/mquellhorst/Workspace/MetaGame/thegame/node_modules/react-dom/cjs/react-dom-server.node.development.js:3373:29)
@metafam/web: [0]     at renderToString (/home/mquellhorst/Workspace/MetaGame/thegame/node_modules/react-dom/cjs/react-dom-server.node.development.js:3988:27)
@metafam/web: [0]     at Object.renderPage (/home/mquellhorst/Workspace/MetaGame/thegame/node_modules/next/dist/next-server/server/render.js:54:854)
@metafam/web: [0]     at Function.getInitialProps (/home/mquellhorst/Workspace/MetaGame/thegame/packages/web/.next/server/pages/_document.js:653:19)
@metafam/web: [0]     at Function.getInitialProps (/home/mquellhorst/Workspace/MetaGame/thegame/packages/web/.next/server/pages/_document.js:1374:85)
@metafam/web: [0]     at loadGetInitialProps (/home/mquellhorst/Workspace/MetaGame/thegame/node_modules/next/dist/next-server/lib/utils.js:5:101)
```

**Comment by richteri** [comment]
According to one the [issue#1361 at draft-js](https://github.com/facebook/draft-js/issues/1361#issuecomment-326763937) you need to include [jsdom](https://www.npmjs.com/package/jsdom) to support SSR by building a custom `stateFromHTML` function.

**Comment by richteri** [comment]
@mquellhorst , I've created a PoC for the jsdom-SSR approach. Please take a look: https://github.com/richteri/TheGame/commit/e05fe96bd0021c2bf8da797193aa9c386bbb5ebf

**Comment by luxumbra** [comment]
@richteri @vidvidvid @Udit-takkar @polimyl are any of you guys up for picking this up and getting it over the finish line? Would be amazing to get it ticked off and the rest of the questing UI sorted.

**Comment by vidvidvid** [comment]
i can work on it after we're done with the dashboard, but if anyone else is willing to pick it up before we're done, go ahead :)

**Comment by Udit-takkar** [comment]
> @richteri @vidvidvid @Udit-takkar @polimyl are any of you guys up for picking this up and getting it over the finish line? Would be amazing to get it ticked off and the rest of the questing UI sorted.

I'll have a look

**Comment by Udit-takkar** [comment]
Tried Creating Quests and Editing it's description afterwards. 
Although there are some minor things left to fix. (any in EditorStateChange, UI dropdown for skills)

Screenshot 2021-11-11 at 12 11 51 PM
Screenshot 2021-11-11 at 12 11 58 PM

**Comment by Udit-takkar** [comment]
Can anyone tell why it is showing so many commits?

**Review by luxumbra — CHANGES_REQUESTED** [review]
Thanks for jumping on this so fast @Udit-takkar 💪🏻 🙏🏻 
Currently when i go to create a Quest it redirects me back to the /quests page.  I am not sure if it was introduced here or in the other commits in this PR but it works on production. 
I *think* I meet the min requirements for creating quests but even if I don't shouldn't there be a notice to let the user know? 
Will check it out locally and take a peek.

**Comment by luxumbra** [comment]
Also, while testing this locally I had a thought that rather than using `[id]` for the quest page we should use `[slug]`...if it's not too much hassle.

**Review by luxumbra — COMMENTED** [review]
Another thing I noticed after copying in some actual content from Notion for Quest 1, the formatting of the HTML in the quest description picks up some css reset on the lists, so probably needs some css added in a `sx` prop on the `<Box>`  on line 95 in `QuestDetails.tsx` or something like that.

**Comment by Udit-takkar** [comment]
> Thanks for jumping on this so fast @Udit-takkar 💪🏻 🙏🏻 Currently when i go to create a Quest it redirects me back to the /quests page. I am not sure if it was introduced here or in the other commits in this PR but it works on production. I _think_ I meet the min requirements for creating quests but even if I don't shouldn't there be a notice to let the user know? Will check it out locally and take a peek.

It shows only on tool tip that you need 100pSeed but not any dialog box when the user clicks. So I think i should be adding a warning when user clicks

**Comment by Udit-takkar** [comment]
Shouldn't this be code in useUser Hook
```
  useEffect(() => {
    if (!redirectTo) return;

    if (
      // If redirectTo is set, redirect if the user was not found.
      redirectTo &&
      redirectIfNotFound &&
      !user
    ) {
      router.push(redirectTo);
    }
  }, [router, user, redirectIfNotFound, redirectTo]);

  return { user, fetching };
};
```

instead of 

```
  useEffect(() => {
    if (!redirectTo) return;

    if (
      // If redirectTo is set, redirect if the user was not found.
      (redirectTo && redirectIfNotFound && !user) ||
      // If redirectIfNotFound is also set, redirect if the user was found
      (redirectIfNotFound && user)
    ) {
      router.push(redirectTo);
    }
  }, [router, user, redirectIfNotFound, redirectTo]);

  return { user, fetching };
};
```

**Comment by luxumbra** [comment]
> redirectTo &&
>       redirectIfNotFound &&
>       !user



>   useEffect(() => {
>     if (!redirectTo) return;
> 
>     if (
>       // If redirectTo is set, redirect if the user was not found.
>       redirectTo &&
>       redirectIfNotFound &&
>       !user
>     ) {
>       router.push(redirectTo);
>     }
>   }, [router, user, redirectIfNotFound, redirectTo]);
> 
>   return { user, fetching };
> };

I have tried this locally and can confirm that it fixes the redirecting issue. If logged in, i can access the create page without it redirecting and when i am disconnected, I am redirected back to /quests. 👍🏻

**Comment by luxumbra** [comment]
@Udit-takkar, I spoke to @peth-yursick earlier and he was saying that we need a way for people who are claiming quests to submit their completed quest. peth suggested that we hook into #did-a-thing channel and the guild onboarding bot to automate this process but would be good to discuss ways of doing this.

**Comment by Udit-takkar** [comment]
> @Udit-takkar, I spoke to @peth-yursick earlier and he was saying that we need a way for people who are claiming quests to submit their completed quest. peth suggested that we hook into #did-a-thing channel and the guild onboarding bot to automate this process but would be good to discuss ways of doing this.

Okay

**Comment by Udit-takkar** [comment]
This branch has been rebased and the Quest Editor has been implemented and whatever the requirement was.
The bot that michiel is working on. Does it has to be included in this PR ?

**Comment by alalonde** [comment]
> @Udit-takkar, I spoke to @peth-yursick earlier and he was saying that we need a way for people who are claiming quests to submit their completed quest. peth suggested that we hook into #did-a-thing channel and the guild onboarding bot to automate this process but would be good to discuss ways of doing this.

Lux, can you add a new issue for this? This PR has been stagnant for too long.

**Comment by alalonde** [comment]
@Udit-takkar Would this be easier if we just rendered the draftjs stuff only on the client side?  The quests pages are not going to be accessed a ton (relative to other pages) so I don't think performance is a huge consideration.

**Comment by Udit-takkar** [comment]
> @Udit-takkar Would this be easier if we just rendered the draftjs stuff only on the client side? The quests pages are not going to be accessed a ton (relative to other pages) so I don't think performance is a huge consideration.

Well this was the blocker in this PR that using draft-js with SSR was giving document undefined . I just fixed this as someone mention this fix https://github.com/facebook/draft-js/issues/1361

**Comment by luxumbra** [comment]
> Lux, can you add a new issue for this? This PR has been stagnant for too long.

Sure. Very good point. 🙏🏻

**Comment by luxumbra** [comment]
> This branch has been rebased and the Quest Editor has been implemented and whatever the requirement was. The bot that michiel is working on. Does it has to be included in this PR ?

No mate. Just me being a bit 'feature creepy'. 😄 I'll add in a new issue

File: packages/web/next.config.js [hunk 1]
```diff
@@ -13,8 +13,15 @@ module.exports = withTM(
         },
       ];
     },
-    future: {
```
Comment (mquellhorst) [inline packages/web/next.config.js:16]: I'm not sure this should be removed.
What's the reason for removing this and isn't this what dysbulic added to use webpack 5 for Glaze?

Comment (Udit-takkar) [reply packages/web/next.config.js:16]: did it by mistake while resolving merge conflicts

Comment (Udit-takkar) [reply packages/web/next.config.js:16]: there is a build error now after adding this 
```
Module not found: Error: Can't resolve 'fs' in 
```

**Comment by Udit-takkar** [comment]
Okay added webpack 5 @mquellhorst .  Do you see anything else wrong?

**Comment by Udit-takkar** [comment]
@luxumbra  rebased it. just try running in your system before merging.

**Comment by luxumbra** [comment]
> @luxumbra rebased it. just try running in your system before merging.

Just checking locally after your rebase. 👍🏻

**Review by luxumbra — APPROVED** [review]
LGTM. Thanks for your work on this @Udit-takkar 🙏🏻

**Review by luxumbra — CHANGES_REQUESTED** [review]
Apologies for the false approval earlier. There's some linting issues that should be resolved before merging this. Please remove or add ignore rules for your console logging if it needs to be there (eg in `design-system`). Also the `any` type warning on line 17 of `storage/handler.ts` and the error on line 120 of `LatestContentTabs/Listen.tsx` needs to be addressed.

**Comment by dysbulic** [comment]
> the `any` type warning on line 17 of `storage/handler.ts` … needs to be addressed

I've got that fixed in my branch.

**Comment by dysbulic** [comment]
> the error on line 120 of `LatestContentTabs/Listen.tsx` needs to be addressed

The error is an unrecognized eslint directive. You can just remove the comment, and it's fine.

**Comment by Udit-takkar** [comment]
Okay thanks, fixing all this in few minues

**Comment by Udit-takkar** [comment]
> ReadableStream



> > the `any` type warning on line 17 of `storage/handler.ts` … needs to be addressed
> 
> I've got that fixed in [my branch](https://github.com/MetaFam/TheGame/pull/943).
> 
> ```
> const tmpFiles = files.map((filename) => ({
>   name: path.basename(filename),
>   stream: () => (
>     (fs.createReadStream(filename) as unknown) as ReadableStream<string>,
>   )
> }));
> ```

It is giving an error
```
Cannot find name 'ReadableStream'.ts(2304)
```

**Review by luxumbra — APPROVED** [review]

**Comment by luxumbra** [comment]
> Okay thanks, fixing all this in few minues

The tests pass now so that is resolved. Thanks. Do we want to wait for [this](https://github.com/MetaFam/TheGame/pull/760#issuecomment-983991546) before it is merged or can we ignore that if it's in @dysbulic branch?

**Comment by dysbulic** [comment]
> It is giving an error: `Cannot find name 'ReadableStream'.ts(2304)`

@Udit-takkar, `ReadableStream` is a part of the standard js library in the browser. It should just work, or at the least be handled by one of the polyfills inserted by Webpack.

@luxumbra, it is safe to leave the `any` for the time being.

**Comment by luxumbra** [comment]
> it is safe to leave the `any` for the time being.

Thanks. Good to know. 🙏🏻

**Pull request merged**